File: test/ModularPipelines.UnitTests/Dependencies/ModuleTagAttributeTests.cs
using System.Reflection;
using ModularPipelines.Attributes;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Dependencies;

public class ModuleTagAttributeTests : TestBase
{
    public async Task Constructor_WithValidTag_SetsTagProperty()
    {
        var attr = new ModuleTagAttribute("database");
        await Assert.That(attr.Tag).IsEqualTo("database");
    }

    public async Task Constructor_WithNullTag_ThrowsArgumentException()
    {
        await Assert.That(() => new ModuleTagAttribute(null!))
            .ThrowsException()
            .And.IsTypeOf<ArgumentException>();
    }

    public async Task Constructor_WithEmptyTag_ThrowsArgumentException()
    {
        await Assert.That(() => new ModuleTagAttribute(""))
            .ThrowsException()
            .And.IsTypeOf<ArgumentException>();
    }

    public async Task Constructor_WithWhitespaceTag_ThrowsArgumentException()
    {
        await Assert.That(() => new ModuleTagAttribute("   "))
            .ThrowsException()
            .And.IsTypeOf<ArgumentException>();
    }

    public async Task Attribute_AllowsMultiple()
    {
        var usage = typeof(ModuleTagAttribute).GetCustomAttribute<AttributeUsageAttribute>();
        await Assert.That(usage!.AllowMultiple).IsTrue();
    }

    public async Task Attribute_IsInheritable()
    {
        var usage = typeof(ModuleTagAttribute).GetCustomAttribute<AttributeUsageAttribute>();
        await Assert.That(usage!.Inherited).IsTrue();
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/ModuleTagAttributeTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open ModularPipelines.UnitTests.Dependencies
open ModularPipelines.UnitTests.FSharp
open TUnit.Core

type ModuleTagAttributeTests() =
    inherit ModularPipelines.UnitTests.Dependencies.ModuleTagAttributeTests()

    [<Test>]
    member this.Constructor_WithValidTag_SetsTagProperty() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleTagAttributeTests> "Constructor_WithValidTag_SetsTagProperty" None

    [<Test>]
    member this.Constructor_WithNullTag_ThrowsArgumentException() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleTagAttributeTests> "Constructor_WithNullTag_ThrowsArgumentException" None

    [<Test>]
    member this.Constructor_WithEmptyTag_ThrowsArgumentException() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleTagAttributeTests> "Constructor_WithEmptyTag_ThrowsArgumentException" None

    [<Test>]
    member this.Constructor_WithWhitespaceTag_ThrowsArgumentException() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleTagAttributeTests> "Constructor_WithWhitespaceTag_ThrowsArgumentException" None

    [<Test>]
    member this.Attribute_AllowsMultiple() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleTagAttributeTests> "Attribute_AllowsMultiple" None

    [<Test>]
    member this.Attribute_IsInheritable() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleTagAttributeTests> "Attribute_IsInheritable" None
